### PR TITLE
fix(no-derived-state): trace values through refs

### PR DESCRIPTION
**Pull request by aidenybai**
## Why

Prop-derived state could evade no-derived-state by passing through useRef.current before reaching a state setter.

## What changed

- Track local useRef initializers and direct current assignments conservatively. Bail out for DOM refs, escaped refs, nested mutation, opaque sources, imports, and external reads.
- Added focused regressions, a patch changeset, and permanent fuzz coverage where this was a confirmed false positive.

## Eval results

The current RDE wrapper cannot consume schema v3 and its rebuilt local path invokes the removed `--diff false` form. I ran the built combined candidate directly over the same first 100 pinned RDE rootDir entries and inspected this rule's filtered results before splitting the identical source change into this PR.

## Test plan

- 69 focused tests; plugin typecheck; 500 strict fuzz iterations; direct candidate-vs-main comparison across 100 pinned RDE root scans added zero unexpected no-derived-state diagnostics; lint/typecheck/format/smoke.

---

> [!NOTE]
> **Medium Risk**
> Changes static analysis in effect value provenance for a core lint rule; behavior is guarded by tests but false positives/negatives are possible on edge-case ref patterns.
> 
> **Overview**
> Closes a **no-derived-state** bypass where props or other render-known values were staged in **`useRef.current`** before an effect called a state setter.
> 
> **`collectValueEvidence`** now follows local **`useRef`** bindings: it merges provenance from the ref’s initializer and simple **`ref.current = …`** writes, and can trace a single **`identifier = …`** reassignment instead of treating any write as opaque. **`ref.current`** reads no longer count as post-mount/external when the binding is a **`useRef`**.
> 
> The analysis stays conservative: compound assignments, **`++`/`--`**, non-**`current`** ref uses, imports, and refs fed from external/DOM-style reads are still treated as unknown so legitimate patterns (layout measurement, externally assigned refs) stay quiet.
> 
> Adds focused **ref provenance** regressions and a patch changeset for **`oxlint-plugin-react-doctor`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f19e6e73da4915b1c7f2e2f3fc55e617da4bc5de. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/utils/collect-effect-state-write-facts.ts [hunk 1]
```diff
@@ -1101,6 +1129,79 @@ const collectValueEvidence = (
...
+            );
+            continue;
+          }
+          if (
+            memberParent &&
+            ((isNodeOfType(memberParent, "AssignmentExpression") && memberParent.left !== member) ||
+              isNodeOfType(memberParent, "UpdateExpression"))
+          ) {
+            evidence.hasUnknownSource = true;
+          }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/utils/collect-effect-state-write-facts.ts:1214]: Fixed in f19e6e73d. Direct ref stores now require operator `=`; compound writes remain unknown. Single-assignment local aliases are resolved from their sole `=` RHS, while multiply-written locals still bail out. Added regressions and reran 72 focused tests, plugin typecheck, and 500 strict fuzz iterations.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/utils/collect-effect-state-write-facts.ts [hunk 2]
```diff
@@ -1101,6 +1129,79 @@ const collectValueEvidence = (
...
+              analysis,
+              initialValue as EsTreeNode,
+              frame,
+              remainingCallFrames,
+              new Set(refVisitedBindings),
+            ),
+          );
+        } else {
+          evidence.hasUnknownSource = true;
+        }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/utils/collect-effect-state-write-facts.ts:1176]: Fixed in f19e6e73d. Bare `useRef()` now contributes the same known initial value as `undefined`, so a later direct prop assignment can establish exclusive render provenance. Added a dedicated regression and reran focused tests/typecheck/fuzz.
